core/local/steps/await_write_finish.js

```diff
@@ -9,95 +8,96 @@ const log = logger({
 
 // Wait this delay (in milliseconds) after the last event for a given file
 // before pushing this event to the next steps.
+// TODO tweak the value (the initial value was chosen because it looks like a
+//      good value, it is not something that was computed)
 const DELAY = 200
 
 /*::
 import type { AtomWatcherEvent } from './event'
 
-type PendingItem = {
-  action: "created" | "modified",
-  fire: *,
-  event: AtomWatcherEvent,
+type WaitingItem = {
+  events: AtomWatcherEvent[],
+  nbCandidates: number,
   timeout: TimeoutID
 }
 */
 
 // TODO add unit tests and logs
 
+function sendReadyBatches (waiting /*: WaitingItem[] */, out /*: Buffer */) {
+  while (waiting.length > 0) {
+    if (waiting[0].nbCandidates !== 0) {
+      break
+    }
+    const item = waiting.shift()
+    clearTimeout(item.timeout)
+    if (item.events.length > 0) {
+      out.push(item.events)
+    }
+  }
+}
+
+function debounce (waiting /*: WaitingItem[] */, events /*: AtomWatcherEvent[] */) {
+  for (let i = 0; i < events.length; i++) {
+    const event = events[i]
+    if (event.kind === 'file' && ['modified', 'deleted'].includes(event.action)) {
+      for (let j = 0; j < waiting.length; j++) {
+        const w = waiting[j]
+        if (w.nbCandidates === 0) { continue }
+        for (let k = 0; k < w.events.length; k++) {
+          const e = w.events[k]
+          if (['created', 'modified'].includes(e.action) && e.path === event.path) {
+            w.events.splice(k, 1)
+            w.nbCandidates--
+            if (event.action === 'modified') {
+              // Preserve the action from the first event (it can be a created file)
+              event.action = e.action
+            }
+            if (event.action === 'deleted' && e.action === 'created') {
+              // It's just a temporary file that we can ignore
+              events.splice(i, 1)
+              i--
+            }
+            break
+          }
+        }
+      }
+    }
+  }
+}
+
 // This is a port of awaitWriteFinish (aWF) from chokidar. It debounces write
 // events for files, as we can have several of them in a short lapse of time,
 // and computing the checksum several times in a row for the same file is not a
 // good idea.
-async function awaitWriteFinish (buffer, out) {
-  const pending = new Map()
+async function awaitWriteFinish (buffer /*: Buffer */, out /*: Buffer */) {
+  const waiting /*: WaitingItem[] */ = []
 
   while (true) {
```

Comment:
How is the loop supposed to stop?
Will something in the `while` block just throw at some point?

Reply:
Would it make sense to separate *stream* stuff (`Buffer`, `while (true)`) from the rest?
Would passing a plain callback (e.g. `out.push`) make sense for lower-level functions?
Wouldn't it make testing edge cases (especially regarding `waiting` events) easier without having to deal with infinite loop + Buffers?

Reply:
Currently, the loop does never stop. I'm not sure if we need to do so or not.

Reply:
About making code easier to test, it's a good idea, but it should probably done when someone will write the test (in another PR)

Reply:
> Currently, the loop does never stop. I'm not sure if we need to do so or not.

Maybe add it to the watcher checklist?